File: backend/src/modules/Portfolio/Portfolio.Domain/Aggregates/Project.ts
import { ICreateProjectParameters } from '../Abstract/ICreateProjectParameters';
import { ProjectMedia } from '../Entities/ProjectMedia';
import { MediaType } from '../Enums/MediaType';
import { PublicationStatus } from '../Enums/PublicationStatus';
import { ProjectCannotBePublishedException } from '../Exceptions/ProjectCannotBePublishedException';
import { ProjectMediaAlreadyAttachedException } from '../Exceptions/ProjectMediaAlreadyAttachedException';
import { ProjectMediaNotFoundException } from '../Exceptions/ProjectMediaNotFoundException';
import { CorporateClientId } from '../ValueObjects/CorporateClientId';
import { DisplayOrder } from '../ValueObjects/DisplayOrder';
import { MediaReference } from '../ValueObjects/MediaReference';
import { ProjectCategoryReference } from '../ValueObjects/ProjectCategoryReference';
import { ProjectId } from '../ValueObjects/ProjectId';
import { ProjectMediaId } from '../ValueObjects/ProjectMediaId';
import { ProjectPeriod } from '../ValueObjects/ProjectPeriod';
import { ProjectServiceReference } from '../ValueObjects/ProjectServiceReference';
import { ProjectTitle } from '../ValueObjects/ProjectTitle';

export class Project {
  private readonly mediaItems: ProjectMedia[] = [];
  private _status = PublicationStatus.Draft;
  private _coverMediaId: ProjectMediaId | null = null;

  private constructor(
    public readonly id: ProjectId,
    private _description: string,
    private _service: ProjectServiceReference,
    private _category: ProjectCategoryReference,
    private _period: ProjectPeriod,
    private _order: DisplayOrder,
    private _title: ProjectTitle | null,
    private _corporateClientId: CorporateClientId | null,
  ) {}

  public static create(input: ICreateProjectParameters): Project {
    return new Project(input.id, input.description ?? '', input.service, input.category, input.period,
      input.order, input.title ?? null, input.corporateClientId ?? null);
  }

  public get title(): ProjectTitle | null { return this._title; }
  public get description(): string { return this._description; }
  public get status(): PublicationStatus { return this._status; }
  public get order(): DisplayOrder { return this._order; }
  public get corporateClientId(): CorporateClientId | null { return this._corporateClientId; }
  public get service(): ProjectServiceReference { return this._service; }
  public get category(): ProjectCategoryReference { return this._category; }
  public get period(): ProjectPeriod { return this._period; }
  public get coverMediaId(): ProjectMediaId | null { return this._coverMediaId; }
  public get media(): readonly ProjectMedia[] { return [...this.mediaItems]; }

  public rename(title: ProjectTitle): void { this._title = title; }
  public changeDescription(description?: string | null): void { this._description = description ?? ''; }
  public publish(): void {
    if (!this._title) throw new ProjectCannotBePublishedException();
    this._status = PublicationStatus.Published;
  }
  public unpublish(): void { this._status = PublicationStatus.Draft; }
  public assignCorporateClient(id: CorporateClientId): void { this._corporateClientId = id; }
  public removeCorporateClient(): void { this._corporateClientId = null; }
  public changeClassification(service: ProjectServiceReference, category: ProjectCategoryReference): void {
    this._service = service;
    this._category = category;
  }
  public changePeriod(period: ProjectPeriod): void { this._period = period; }
  public changeOrder(order: DisplayOrder): void { this._order = order; }

  public addMedia(id: ProjectMediaId, reference: MediaReference, type: MediaType, order: DisplayOrder): ProjectMedia {
    if (this.mediaItems.some((item) => item.id.equals(id))) {
      throw new ProjectMediaAlreadyAttachedException(id.value);
    }
    const media = new ProjectMedia(id, reference, type, order);
    this.mediaItems.push(media);
    return media;
  }

  public updateMedia(id: ProjectMediaId, reference: MediaReference, type: MediaType, order: DisplayOrder): void {
    const index = this.findMediaIndex(id);
    this.mediaItems[index] = this.mediaItems[index]!.withChanges(reference, type, order);
  }

  public changeMediaOrder(id: ProjectMediaId, order: DisplayOrder): void {
    const index = this.findMediaIndex(id);
    this.mediaItems[index] = this.mediaItems[index]!.withOrder(order);
  }

  public setCoverMedia(id: ProjectMediaId): void { this.findMediaIndex(id); this._coverMediaId = id; }
  public clearCoverMedia(): void { this._coverMediaId = null; }
  public removeMedia(id: ProjectMediaId): void {
    this.mediaItems.splice(this.findMediaIndex(id), 1);
    if (this._coverMediaId?.equals(id)) this._coverMediaId = null;
  }

  private findMediaIndex(id: ProjectMediaId): number {
    const index = this.mediaItems.findIndex((item) => item.id.equals(id));
    if (index < 0) throw new ProjectMediaNotFoundException(id.value);
    return index;
  }
}

File: backend/modules/Portfolio/CromaticaCreativa.Modules.Portfolio.Domain/Aggregates/Project.cs
using CromaticaCreativa.Modules.Portfolio.Domain.Entities;
using CromaticaCreativa.Modules.Portfolio.Domain.Enums;
using CromaticaCreativa.Modules.Portfolio.Domain.Exceptions;
using CromaticaCreativa.Modules.Portfolio.Domain.ValueObjects;

namespace CromaticaCreativa.Modules.Portfolio.Domain.Aggregates;

public sealed class Project
{
    private readonly List<ProjectMedia> _media = [];

    private Project(
        ProjectId id,
        string? description,
        ProjectServiceReference service,
        ProjectCategoryReference category,
        ProjectPeriod period,
        DisplayOrder order,
        ProjectTitle? title,
        CorporateClientId? corporateClientId)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        Description = description ?? string.Empty;
        Service = service ?? throw new ArgumentNullException(nameof(service));
        Category = category ?? throw new ArgumentNullException(nameof(category));
        Period = period ?? throw new ArgumentNullException(nameof(period));
        Order = order;
        Title = title;
        CorporateClientId = corporateClientId;
        Status = PublicationStatus.Draft;
    }

    public ProjectId Id { get; }

    public ProjectTitle? Title { get; private set; }

    public string Description { get; private set; }

    public PublicationStatus Status { get; private set; }

    public DisplayOrder Order { get; private set; }

    public CorporateClientId? CorporateClientId { get; private set; }

    public ProjectServiceReference Service { get; private set; }

    public ProjectCategoryReference Category { get; private set; }

    public ProjectPeriod Period { get; private set; }

    public ProjectMediaId? CoverMediaId { get; private set; }

    public IReadOnlyCollection<ProjectMedia> Media => _media.AsReadOnly();

    public static Project Create(
        ProjectId id,
        string? description,
        ProjectServiceReference service,
        ProjectCategoryReference category,
        ProjectPeriod period,
        DisplayOrder order,
        ProjectTitle? title = null,
        CorporateClientId? corporateClientId = null) =>
        new(id, description, service, category, period, order, title, corporateClientId);

    public void Rename(ProjectTitle title) =>
        Title = title ?? throw new ArgumentNullException(nameof(title));

    public void ChangeDescription(string? description) => Description = description ?? string.Empty;

    public void Publish()
    {
        if (Title is null)
        {
            throw new ProjectCannotBePublishedException();
        }

        Status = PublicationStatus.Published;
    }

    public void Unpublish() => Status = PublicationStatus.Draft;

    public void AssignCorporateClient(CorporateClientId corporateClientId) =>
        CorporateClientId = corporateClientId ?? throw new ArgumentNullException(nameof(corporateClientId));

    public void RemoveCorporateClient() => CorporateClientId = null;

    public void ChangeClassification(
        ProjectServiceReference service,
        ProjectCategoryReference category)
    {
        Service = service ?? throw new ArgumentNullException(nameof(service));
        Category = category ?? throw new ArgumentNullException(nameof(category));
    }

    public void ChangePeriod(ProjectPeriod period) =>
        Period = period ?? throw new ArgumentNullException(nameof(period));

    public void ChangeOrder(DisplayOrder order) => Order = order;

    public ProjectMedia AddMedia(
        ProjectMediaId mediaId,
        MediaReference reference,
        MediaType type,
        DisplayOrder order)
    {
        ArgumentNullException.ThrowIfNull(mediaId);

        if (_media.Any(item => item.Id == mediaId))
        {
            throw new ArgumentException($"Project media '{mediaId.Value}' is already attached.", nameof(mediaId));
        }

        var media = new ProjectMedia(mediaId, reference, type, order);
        _media.Add(media);
        return media;
    }

    public void UpdateMedia(
        ProjectMediaId mediaId,
        MediaReference reference,
        MediaType type,
        DisplayOrder order)
    {
        var media = FindMedia(mediaId);
        media.ChangeReference(reference);
        media.ChangeType(type);
        media.ChangeOrder(order);
    }

    public void ChangeMediaOrder(ProjectMediaId mediaId, DisplayOrder order) =>
        FindMedia(mediaId).ChangeOrder(order);

    public void SetCoverMedia(ProjectMediaId mediaId)
    {
        FindMedia(mediaId);
        CoverMediaId = mediaId;
    }

    public void ClearCoverMedia() => CoverMediaId = null;

    public void RemoveMedia(ProjectMediaId mediaId)
    {
        var media = FindMedia(mediaId);
        _media.Remove(media);

        if (CoverMediaId == mediaId)
        {
            CoverMediaId = null;
        }
    }

    private ProjectMedia FindMedia(ProjectMediaId mediaId)
    {
        ArgumentNullException.ThrowIfNull(mediaId);
        return _media.FirstOrDefault(item => item.Id == mediaId)
            ?? throw new ProjectMediaNotFoundException(mediaId);
    }
}
